### ISSUE TITLE
String.split hangs when string ends with quotation mark (QU class)

### ISSUE THREAD

**Issue by GregLMcDonald** [elixir-unicode/unicode_string#5]
I've found that with the `break: :line` option, `String.split/2` hangs if the string argument ends with any quotation mark character. All the following will hang:

```
Unicode.String.split(~s(He said, "A cup of hot tea?"), locale: :en, break: :line) # Quotation mark (aka typewriter quote)
Unicode.String.split(~s(He said, “A cup of hot tea?”), locale: :en, break: :line) # Right double quotation mark
Unicode.String.split(~s(«Voilà le problème!»), locale: :fr, break: :line) # Right-pointing double angle quotation mark
```

I would expect `Unicode.String.split(~s(He said, "A cup of hot tea?"), locale: :en, break: :line)` to return `["He ", "said, ", "\"A ", "cup ", "of ", "hot ", "tea?\""]` and this is indeed how the online [Unicode segmentation utility](https://util.unicode.org/UnicodeJsps/breaks.jsp) splits the same text. The same text is not problematic if `break: :word` is used. My test are with version 1.3 of the library. Elixir 1.14.

I poked around the code a bit and the Unicode line segmentation algo but must admit I find it all rather daunting. In the line segmentation test data, it doesn't _seem_ like there's anything that would prevent a break after a quotation mark.

As a workaround, for now I am padding the end of the string with a single space. This allows the method to return and for my use case is perfectly fine.

Let me take this opportunity to thank you @kipcole9 for this incredible library and all your contributions to the Elixir ecosystem.

**Comment by kipcole9** [comment]
thanks for the kind words @GregLMcDonald, and apologies for the inconvenience. Sounds like I've got an infinite recursion bug I need to chase down - which I will do now.

Yes, the Unicode and CLDR worlds are vast. And every time I think there's a small project I can experiment with it becomes a large project. But ultimately good fun and great learning.

**Comment by kipcole9** [comment]
BTW, Unicode Line segmentation, Word segmentation and Grapheme Cluster segmentation are three very different algorithms.  So no particular surprise the results differ.

I am coincidentally working on a dictionary-base word segmented for Chinese, Japanese, Lao, Burmese and Thai to better match what the ICU utilities spit out. So thankfully my brain is already in the right mindset for this.

**Comment by kipcole9** [comment]
I believe this is fixed now. Would you consider configuring `unicode_string` from GitHub and testing if it resolves the issue to your satisfaction?
```elixir
{:unicode_string, github: "elixir-unicode/unicode_string"},
```

When its ok from your end I'll publish to `hex.pm`.

**Comment by GregLMcDonald** [comment]
Thanks for the amazing responsiveness. This fixes the issue with the ending quotation mark but it introduces another interesting bug.

If `string` in `Unicode.String.split(string, locale: <correct locale>, break: :word)` contains a quotation mark and ends without any additional word break opportunities, the quotation mark is generally repeated. Here are some examples of this behaviour. It is not present in v1.3, only when I point to a312929625b53f0664ab0c9d90782db241678bca.

```
~s("Du) |> Unicode.String.split(locale: :de, break: :word)
["\"", "\"Du"]

~s(“Hi) |> Unicode.String.split(locale: :en, break: :word)
["“", "“Hi"]
```

When I tried some `:fr` cases, I found that `~s(«Du)` reproduced the bug but oddly many other strings didn't, eg. `~s(«Oui)`.

If a "closing" quotation mark or any break opportunity (like a space) is added, everything is fine.

```
~s("Hi") |> Unicode.String.split(locale: :en, break: :word)
["\"", "Hi", "\""]

~s("Hi ) |> Unicode.String.split(locale: :en, break: :word)
["\"", "Hi", " "]
```

**Comment by kipcole9** [comment]
> If a "closing" quotation mark or any break opportunity (like a space) is added, everything is fine.

Yes, it's the same issue as the original case you opened. For which my original fix wasn't correct. I pushed another [commit](https://github.com/elixir-unicode/unicode_string/commit/2ae3939f635639957dbb507d333d92d38d0dd5cd) which I believe fixes the fix (although I've said that before!).

I think everything is now as it should be but I'd appreciate your confirmation and feedback.

**Comment by GregLMcDonald** [comment]
That seems to have done it. Thanks again.

Thought you might be interested to see the feature I'm developing that uses `split/2`. I need to wrap the words in a text in `<button>` tags (for a "word reader" feature) while preserving proper line breaks by wrapping unbreakable sequences in `<span>` styled with `white-space: nowrap`. The video didn't capture the audio (while the word is unlerlined, the word is read aloud):

https://github.com/elixir-unicode/unicode_string/assets/8013211/ebe3fbee-3361-4711-9e1a-5599b68af851

**Comment by kipcole9** [comment]
Thats fascinating - thanks very much for sharing. Does this eventually end up in a public web site someone?

Please do send along any feature requests that would make you life easier for this kind of work.

Depending on how big your text is, you might also find `Unicode.String.next/2` useful since it is kind of a streaming break iterator.  `Unicode.String.stream/2` is a full streaming interface built upon `Unicode.String.next/2`.

**Comment by kipcole9** [comment]
I've published [unicode_string version 1.3.1](https://hex.pm/packages/unicode_string/1.3.1) with the following changelog entry:

### Bug Fixes

* Fix `Unicode.String.split/2` and `Unicode.String.next/2` when the passing rule is `:no_break` rule. Thanks to @GregLMcDonald for the report. Closes #5.

**Comment by GregLMcDonald** [comment]
The streaming methods are a great addition. For now, I'm doing a batch transformation on the fly for a LiveView. The texts are short, so it's not a problem.

The feature is the "virtual reading room" part of https://readersadvance.com which is in closed beta. I'd love to invite you to the platform to have a look around. Just click the "Be a Beta Tester" button on the landing page and you can submit your email address through the form.
